okta/config.go

```diff
@@ -47,12 +47,12 @@ type config struct {
 				MaxRetries int32 `yaml:"maxRetries" envconfig:"OKTA_CLIENT_RATE_LIMIT_MAX_RETRIES"`
 				MaxBackoff int64 `yaml:"maxBackoff" envconfig:"OKTA_CLIENT_RATE_LIMIT_MAX_BACKOFF"`
 			} `yaml:"rateLimit"`
-			OrgUrl            string   `yaml:"orgUrl" envconfig:"OKTA_CLIENT_ORGURL"`
-			Token             string   `yaml:"token" envconfig:"OKTA_CLIENT_TOKEN"`
-			AuthorizationMode string   `yaml:"authorizationMode" envconfig:"OKTA_CLIENT_AUTHORIZATIONMODE"`
-			ClientId          string   `yaml:"clientId" envconfig:"OKTA_CLIENT_CLIENTID"`
-			Scopes            []string `yaml:"scopes" envconfig:"OKTA_CLIENT_SCOPES"`
-			PrivateKey        string   `yaml:"privateKey" envconfig:"OKTA_CLIENT_PRIVATEKEY"`
+			OrgUrl            string      `yaml:"orgUrl" envconfig:"OKTA_CLIENT_ORGURL"`
+			Token             string      `yaml:"token" envconfig:"OKTA_CLIENT_TOKEN"`
+			AuthorizationMode string      `yaml:"authorizationMode" envconfig:"OKTA_CLIENT_AUTHORIZATIONMODE"`
+			ClientId          string      `yaml:"clientId" envconfig:"OKTA_CLIENT_CLIENTID"`
+			Scopes            []string    `yaml:"scopes" envconfig:"OKTA_CLIENT_SCOPES"`
+			PrivateKey        interface{} `yaml:"privateKey" envconfig:"OKTA_CLIENT_PRIVATEKEY"`
```

Comment:
Using an `interface` type would mean a lot of type checking / conversions, is there a more concrete type that could be used here?

Reply:
The usage of `interface{}` was outlined as one of the considerations for my approach in the PR description. Currently, the SDK uses the config's `PrivateKey` field on each new request, not when the client is initialized. This usage of type assertions does not have a severe impact on performance for a few reasons:

- The type assertion will only occur in three situations: once on new requests that skip the cache (approx one per hour); and twice when the client is initialized + validated.
- Type assertion in Go has been quite efficient since v1.11 with benchmarks around ~2 ns/op.

Using an `interface{}` means there are minimal changes to the RequestExecutor code and allows the configuration to support multi-typed values.

Reply:
If the maintainers are open to modifying the `RequestExecutor.NewRequest()` code, the following changes could be made to remove the `interface{}` and increase performance further:

1. Update the `config` struct to include a `PrivateKeySigner` attribute of type `jose.Signer`. This can be accompanied by a `okta.WithPrivateKeySigner(jose.Signer)` method to match the functionality of `okta.WithHttpClientPtr(*http.Client)`.

```go
type config struct {
	PrivateKeySigner jose.Signer
}

func WithPrivateKeySigner(signer jose.Signer) ConfigSetter {
	return func(c *config) {
		c.PrivateKeySigner = signer
	}
}
```

2. Add a `privateKeySigner` attribute to the `RequestExecutor` struct.

```go
type RequestExecutor struct {
	...
	privateKeySigner  jose.Signer
}
```

3. Update the `NewRequestExecutor(...)` return signature to include an `error`. This will allow the private key parsing code into the constructor and occur at the time of the client initialization.

```go
func NewRequestExecutor(httpClient *http.Client, cache cache.Cache, config *config) (*RequestExecutor, error) {
	...
	if config.Okta.Client.AuthorizationMode == "PrivateKey" {
		if config.PrivateKeySigner != nil {
			re.privateKeySigner = config.PrivateKeySigner
		} else {
			priv := []byte(strings.ReplaceAll(config.Okta.Client.PrivateKey, `\n`, "\n"))

			privPem, _ := pem.Decode(priv)
			if privPem == nil {
				return nil, errors.New("invalid private key")
			}
			if privPem.Type != "RSA PRIVATE KEY" {
				return nil, fmt.Errorf("RSA private key is of the wrong type")
			}

			parsedKey, err := x509.ParsePKCS1PrivateKey(privPem.Bytes)
			if err != nil {
				return nil, err
			}

			if re.privateKeySigner, err = jose.NewSigner(jose.SigningKey{Algorithm: jose.RS256, Key: parsedKey}, nil); err != nil {
				return nil, err
			}
		}
	}

	return &re, nil
}
```

4. Update the client validation logic to allow an empty `config.Okta.Client.PrivateKey` if `config.PrivateKeySigner` is not `nil`.

```go
func validateAuthorization(c *config) error {
	...

	if c.Okta.Client.AuthorizationMode == "PrivateKey" &&
		(c.Okta.Client.ClientId == "" ||
			c.Okta.Client.Scopes == nil ||
			(c.Okta.Client.PrivateKey == "" &&
				c.PrivateKeySigner == nil)) {
		return errors.New("when using AuthorizationMode 'PrivateKey', you must supply 'ClientId', 'Scopes', and 'PrivateKey'")
	}

	return nil
}
```

5. Change the Okta client constructor to support the modified `NewRequestExecutor()` signature.

```go
func NewClient(ctx context.Context, conf ...ConfigSetter) (context.Context, *Client, error) {
	...
	c := &Client{}
	c.config = config

	c.requestExecutor, err = NewRequestExecutor(config.HttpClient, oktaCache, config)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to initialize request executor: %w", err)
	}

	c.resource.client = c
	...
}
```

These changes will increase performance overall as the private key validation and signer construction will only occur at when the Okta client is initialized. Then, the subsequent usage within the `RequestExecutor.NewRequest()` method can be modified to simply utilize the pre-constructed signer.

```go
func (re *RequestExecutor) NewRequest(method string, url string, body interface{}) (*http.Request, error) {
	...
	if re.config.Okta.Client.AuthorizationMode == "PrivateKey" {
		if re.cache.Has(AccessTokenCacheKey) {
			token := re.cache.GetString(AccessTokenCacheKey)
			req.Header.Add("Authorization", "Bearer "+token)
		} else {
			claims := ClientAssertionClaims{
				...
			}
			jwtBuilder := jwt.Signed(re.privateKeySigner).Claims(claims)
			clientAssertion, err := jwtBuilder.CompactSerialize()
			if err != nil {
				return nil, err
			}
			...
		}
	}
	...
}
```

This approach also has a couple of considerations, such as the inability to modify the private key between requests, and there are additional signature changes. Happy to proceed with this approach after the maintainers have weighed in.